Add caching for uploaded header images

pbearne opened the pull request:
Previously, the `get_uploaded_header_images` function did not utilize caching, leading to potential performance issues. This commit introduces transient caching to store the header images, significantly improving efficiency. A corresponding PHPUnit test has also been added to verify the caching functionality.


Trac ticket: https://core.trac.wordpress.org/ticket/49446

Commit:
Add caching for uploaded header images

Previously, the `get_uploaded_header_images` function did not utilize caching, leading to potential performance issues. This commit introduces transient caching to store the header images, significantly improving efficiency. A corresponding PHPUnit test has also been added to verify the caching functionality.

Commit:
Fix dynamic IDs in getUploadedHeaderImages test

This commit updates the getUploadedHeaderImages test to use dynamic IDs for attachments and parents instead of hardcoded values. This change ensures greater flexibility and accuracy in testing scenarios by reflecting actual usage.

mukeshpanchal27 commented:
@pbearne some unit tests are failing. Could you please take a look?

pbearne committed:
Merge branch 'WordPress:trunk' into 49446-cache-the-results-of-get_uploaded_header_images

mukeshpanchal27 commented on src/wp-includes/theme.php:
```diff
@@ -1498,14 +1498,22 @@ function header_image() {
  * @return array
  */
 function get_uploaded_header_images() {
+
```
```suggestion
```

mukeshpanchal27 commented:
@SergeyBiryukov or @desrosj could one of you please take a look when you have moement. Thanks!

felixarntz commented on src/wp-includes/theme.php:
```diff
@@ -1540,6 +1548,8 @@ function get_uploaded_header_images() {
 		}
 	}
 
+	set_transient( $transient_key, $header_images, HOUR_IN_SECONDS );
```
I think this is something that would be reasonable to cache for longer than 1 hour, e.g. 1 day at least.

If the theme is switched, this will automatically be invalidated anyway, and the other crucial piece (see my overarching review feedback) is that we'll need to invalidate when such an image is newly uploaded.

pbearne replied:
set top a day

felixarntz reviewed (changes requested):
@pbearne This looks solid in terms of caching logic. However, we also need to make sure that this is invalid when relevant attachments are uploaded/assigned/modified.

Can you implement additional hook callbacks here that delete the transient whenever an attachment is created or assigned to the active theme (via the `_wp_attachment_is_custom_header` meta key/value), or when such a an attachment is edited?

pbearne committed:
Update src/wp-includes/theme.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

Commit:
Add cache invalidation for header images

Introduce `invalidate_header_images_cache` function to clear cached header images when custom header meta is updated or deleted. Add corresponding unit tests to ensure cache invalidation logic works correctly.

pbearne commented:
> @pbearne This looks solid in terms of caching logic. However, we also need to make sure that this is invalid when relevant attachments are uploaded/assigned/modified.
> 
> Can you implement additional hook callbacks here that delete the transient whenever an attachment is created or assigned to the active theme (via the `_wp_attachment_is_custom_header` meta key/value), or when such a an attachment is edited?

@felixarntz
Have add cache invalidation for the meta I belive that this will also cover if it is edited as this meta will be updated as part of the process

felixarntz commented on src/wp-includes/theme.php:
```diff
@@ -1540,9 +1547,33 @@ function get_uploaded_header_images() {
+		delete_transient( $transient_key );
+	}
+}
+add_action( 'updated_post_meta', 'invalidate_header_images_cache', 10, 4 );
+add_action( 'deleted_post_meta', 'invalidate_header_images_cache', 10, 4 );
```
Please include `added_post_meta`, which is currently missing, but could affect this cache in the same way.
```suggestion
add_action( 'added_post_meta', 'invalidate_header_images_cache', 10, 4 );
add_action( 'updated_post_meta', 'invalidate_header_images_cache', 10, 4 );
add_action( 'deleted_post_meta', 'invalidate_header_images_cache', 10, 4 );
```

felixarntz commented on src/wp-includes/theme.php:
```diff
@@ -1540,9 +1547,33 @@ function get_uploaded_header_images() {
+/**
+ * Invalidates the cache for header images when a custom header image is changed or deleted.
+ *
+ * @since 6.7
```
```suggestion
 * @since 6.7.0
```

felixarntz commented on src/wp-includes/theme.php:
```diff
@@ -1498,14 +1498,21 @@ function header_image() {
  */
 function get_uploaded_header_images() {
+	$stylesheet    = get_option( 'stylesheet' );
+	$transient_key = 'uploaded_header_images' . $stylesheet;
```
A bit of a weird key because there's no underscore between the word "images" and the theme slug, let's fix that:
```suggestion
	$transient_key = "uploaded_header_images_{$stylesheet}";
```

felixarntz commented on src/wp-includes/theme.php:
```diff
@@ -1540,9 +1547,33 @@ function get_uploaded_header_images() {
+function invalidate_header_images_cache( $meta_id, $post_id, $meta_key, $meta_value ) {
+	if ( '_wp_attachment_is_custom_header' === $meta_key ) {
+		$stylesheet    = get_option( 'stylesheet' );
+		$transient_key = 'uploaded_header_images' . $stylesheet;
```
See above:
```suggestion
		$transient_key = "uploaded_header_images_{$stylesheet}";
```

felixarntz commented on src/wp-includes/theme.php:
```diff
@@ -1540,9 +1547,33 @@ function get_uploaded_header_images() {
+ */
+function invalidate_header_images_cache( $meta_id, $post_id, $meta_key, $meta_value ) {
+	if ( '_wp_attachment_is_custom_header' === $meta_key ) {
+		$stylesheet    = get_option( 'stylesheet' );
```
This needs to respect `$meta_value`. In most cases, it'll be the same as the current stylesheet, but it could also be updated for another theme than the current theme - unlikely, but possible. At the moment you aren't using `$meta_value` at all.

Note that we also need to still cater for the case that `$meta_value` is empty (e.g. upon deletion), so we should use the current stylesheet only as a fallback for that case.
```suggestion
		$stylesheet    = $meta_value ? $meta_value : get_option( 'stylesheet' );
```

felixarntz commented on tests/phpunit/tests/theme/getUploadedHeaderImages.php:
```diff
@@ -0,0 +1,62 @@
+
+		$this->assertSame( $expected, get_uploaded_header_images() );
+
+		$this->assertSame( $num_queries, get_num_queries() );
```
This assertion is currently failing, see https://github.com/WordPress/wordpress-develop/actions/runs/10839289258/job/30079360953?pr=7154.

felixarntz commented on tests/phpunit/tests/theme/invalidateHeaderImagesCache.php:
```diff
@@ -0,0 +1,112 @@
+
+		$this->assertSame( $expected, get_uploaded_header_images() );
+
+		$this->assertSame( $num_queries, get_num_queries() );
```
See above, this is failing in https://github.com/WordPress/wordpress-develop/actions/runs/10839289258/job/30079360953?pr=7154.

felixarntz commented on tests/phpunit/tests/theme/invalidateHeaderImagesCache.php:
```diff
@@ -0,0 +1,112 @@
+
+		$this->assertSame( $expected, get_uploaded_header_images() );
+
+		$this->assertSame( $num_queries, get_num_queries() );
```
See https://github.com/WordPress/wordpress-develop/actions/runs/10839289258/job/30079360953?pr=7154

felixarntz reviewed (commented):
@pbearne 
> Have add cache invalidation for the meta I belive that this will also cover if it is edited as this meta will be updated as part of the process

The invalidation you've added is not yet reliable. It needs to also cater for when the image is edited - that won't happen automatically. The cached data in `get_uploaded_header_images()` includes things like the image URL, alt text etc. If those change, we need to invalidate the new cache here. So you need to include a few more actions, e.g. when updating the attachment, or when updating any metadata on the attachment that is currently set as `_wp_attachment_is_custom_header`.

pbearne committed:
Update src/wp-includes/theme.php

Co-authored-by: Felix Arntz <flixos90@gmail.com>

pbearne committed:
Update src/wp-includes/theme.php

Co-authored-by: Felix Arntz <flixos90@gmail.com>

pbearne committed:
Update src/wp-includes/theme.php

Co-authored-by: Felix Arntz <flixos90@gmail.com>

pbearne committed:
Update src/wp-includes/theme.php

Co-authored-by: Felix Arntz <flixos90@gmail.com>

pbearne committed:
Update src/wp-includes/theme.php

Co-authored-by: Felix Arntz <flixos90@gmail.com>

felixarntz commented:
@pbearne Given that 6.7 Beta 1 is tomorrow, we'll have to punt this to 6.8. There's still two outstanding issues here:
* The cache invalidation needs to cater for when the attachment that is currently set as `_wp_attachment_is_custom_header` gets edited. See https://github.com/WordPress/wordpress-develop/pull/7154#pullrequestreview-2323622986.
* The tests are failing when using a persistent object cache.

pbearne committed:
Merge branch 'WordPress:trunk' into 49446-cache-the-results-of-get_uploaded_header_images

Commit:
Update PHPDoc @covers tag in theme tests

Replaced @coversDefaultClass with @covers for better specificity in theme invalidateHeaderImagesCache test. This adjustment enhances the precision of test coverage annotation and maintains the accuracy of test documentation. No actual code or functionality was altered.

felixarntz commented:
@pbearne Some of the PHPUnit tests are still failing, e.g. https://github.com/WordPress/wordpress-develop/actions/runs/11507618643/job/32034197173?pr=7154.

Would you be able to take another look at addressing those? Please let us know if you run into any blockers.

swissspidy committed:
Merge branch 'trunk' into 49446-cache-the-results-of-get_uploaded_header_images

Commit:
Adjust header image query count based on object cache usage in theme tests

Commit:
Standardize spacing in conditional statements within theme test files

westonruter commented on tests/phpunit/tests/theme/getUploadedHeaderImages.php:
```diff
@@ -0,0 +1,66 @@
+			$cropped_1_id => array(
+				'attachment_id'     => $cropped_1_id,
+				'url'               => 'http://example.org/wp-content/uploads/foo-cropped-1.png',
+				'thumbnail_url'     => 'http://example.org/wp-content/uploads/foo-cropped-1.png',
```
The tests are failing when `example.org` is used. The actual host in the URLs is `example.org:8889`. So the host name here can't be hard-coded in this test or the other tests introduced here.

westonruter commented on src/wp-includes/theme.php:
```diff
@@ -1498,14 +1498,21 @@ function header_image() {
+
 	$header_images = array();
 
-	// @todo Caching.
```
Question: What about using the `cache_results` query var in the call to `get_posts()`? It's actually defaulting to `true`. This will cache the results as well, although in object cache only. So sites with a persistent object cache will benefit, but others will not. Still, because the transient has an expiration, it won't be auto-loaded, so there will be a database query still regardless, right?

pbearne replied:
If we just relied on the get_posts as you said only site with persistent object cache will benefit
With the transient all sites will benefit (and the transients maybe in the persistent object cache) 
This helps more site like this I belive

westonruter commented on src/wp-includes/theme.php:
```diff
@@ -1498,14 +1498,21 @@ function header_image() {
  * @return array
  */
 function get_uploaded_header_images() {
+	$stylesheet    = get_option( 'stylesheet' );
```
Why not `get_stylesheet()`?


```suggestion
	$stylesheet    = get_stylesheet();
```

pbearne committed:
Update src/wp-includes/theme.php

Co-authored-by: Weston Ruter <westonruter@gmail.com>

pbearne committed:
Merge branch 'WordPress:trunk' into 49446-cache-the-results-of-get_uploaded_header_images

Commit:
Replace hardcoded URLs with dynamic `get_site_url()` calls in theme tests. This ensures greater flexibility and alignment with site configurations during test execution.
